File: speed_test/cpp/for_loops.cpp
#include <iostream>
#include <ctime>
#include <array>
using namespace std;

int main()
{
  clock_t time_req; // clock object
  unsigned long long int outerloop = 10000; // outer loop size
  unsigned long long int innerloop = 10000; // inner loop size
  unsigned long long int num_experiments = 10; // number of experiments
  float total_times[num_experiments]; // array with time for each experiment

  cout << outerloop << "x" << innerloop << " Nested For Loops" << endl;

  // generate random matrix and invert it
  for(unsigned long long int i=0; i<num_experiments; i++)
  {
    cout << "Experiment " << i+1 << " of " << num_experiments;
    unsigned long long int counter = 0;
    time_req = clock();
    for(unsigned long long int j=0; j<outerloop; j++)
    {
      unsigned long long int counter2 = 0;
      for(unsigned long long int k=0; k<innerloop; k++)
      {
        counter2+=k;
      }
      counter+=counter2;
    }
    time_req = clock() - time_req; // time inversion only
    total_times[i] = (float)time_req/CLOCKS_PER_SEC;
    cout << "     Took: " << total_times[i] << "seconds" << endl;
    cout<<"counter "<< counter<<endl;
  }

  float avrg = 0;
  for(unsigned long long int i=0; i<num_experiments; i++)
  {
    avrg+=total_times[i];
  }
  cout << "~~~~~~~~~~~~~~~" << endl;
  cout << "Ran "<< num_experiments << " experiments in: " << avrg << endl;
  avrg/=num_experiments;
  cout << "Average: "<< avrg << endl;
  return 0;
}

File: speed_test/matlab/for_loops.m
% Script that runs nested for loops
outerloop = 10000; % outer loop size
innerloop = 10000; % inner loop size
num_experiments = 10; % number of experiments
total_times = zeros(num_experiments,1);

msg = strcat(num2str(outerloop), "x", num2str(innerloop)," Nested For Loops");
disp(msg)

% run nested loops
for  i =1: num_experiments
    counter = 0;
    tic;
    for j =0:outerloop-1
        counter2 = 0;
        for k =0:innerloop-1
            counter2=counter2+k;
        end
        counter=counter+counter2;
    end
    t = toc;
    msg = strcat("Nester Loops  ", num2str(i+1), " of ", strcat(num_experiments), "    Took: ", num2str(t), " seconds");
    disp(msg)
    total_times(i) = t;
    disp(counter)
end

total_time = sum(total_times);
avrg = total_time/num_experiments;

disp("~~~~~~~~~~~~~~~")
msg = strcat("Ran ", num2str(num_experiments), " experiments in: ", num2str(total_time));
disp(msg)
msg = strcat("Average: ", num2str(avrg));
disp(msg)
